video/out/vo_gpu_next.c

```diff
@@ -1515,84 +1521,216 @@ static void wait_events(struct vo *vo, int64_t until_time_ns)
-done:
-    talloc_free(dir);
+static int compare_atime(const void *a, const void *b)
+{
+    return (((struct file_entry *)b)->atime - ((struct file_entry *)a)->atime);
 }
 
 static void cache_uninit(struct priv *p, struct cache *cache)
 {
     if (!cache->cache)
-        goto done;
-    if (pl_cache_signature(cache->cache) == cache->sig)
-        goto done; // skip re-saving identical cache
+        return;
 
-    assert(cache->path);
-    char *tmp = talloc_asprintf(cache->path, "%sXXXXXX", cache->path);
-    int fd = mkstemp(tmp);
-    if (fd < 0)
-        goto done;
-    FILE *file = fdopen(fd, "wb");
-    if (!file) {
-        close(fd);
-        unlink(tmp);
+    void *ta_ctx = talloc_new(NULL);
+    struct file_entry *files = talloc_array(ta_ctx, struct file_entry, 0);
+    size_t num_files = 0;
+    assert(cache->dir);
+    assert(cache->name);
+
+    DIR *d = opendir(cache->dir);
+    if (!d)
         goto done;
+
+    struct dirent *dir;
+    while ((dir = readdir(d)) != NULL) {
+        char *filepath = mp_path_join(ta_ctx, cache->dir, dir->d_name);
+        if (!filepath)
+            continue;
+        struct stat filestat;
+        if (stat(filepath, &filestat))
+            continue;
+        if (!S_ISREG(filestat.st_mode))
+            continue;
+        bstr fname = bstr0(dir->d_name);
+        if (!bstr_eatstart0(&fname, cache->name))
+            continue;
+        if (fname.len != 17) // _%016x
+            continue;
+        MP_TARRAY_APPEND(ta_ctx, files, num_files,
+                         (struct file_entry){
+                             .filepath = filepath,
+                             .size     = filestat.st_size,
+                             .atime    = filestat.st_atime,
+                         });
     }
-    int ret = pl_cache_save_file(cache->cache, file);
-    fclose(file);
-    if (ret >= 0)
-        ret = rename(tmp, cache->path);
-    if (ret < 0) {
-        MP_WARN(p, "Failed saving cache to %s\n", cache->path);
-        unlink(tmp);
+    closedir(d);
+
+    if (!num_files)
+        goto done;
+
+    qsort(files, num_files, sizeof(struct file_entry), compare_atime);
+
+    time_t t = time(NULL);
+    MP_DBG(p, "Cleaning %s cache directory", cache->name);
+    if (cache->size_limit)
+        MP_DBG(p, " (limit: %zu B)", cache->size_limit);
+    MP_DBG(p, ":\n");
+    size_t cache_size = 0;
+    size_t cache_limit = cache->size_limit ? cache->size_limit : SIZE_MAX;
+    for (int i = 0; i < num_files; i++) {
+        // Remove files that exceed the size limit but are older than one day.
+        // This allows for temporary maintaining a larger cache size while
+        // adjusting the configuration. The cache will be cleared the next day
+        // for unused entries. We don't need to be overly aggressive with cache
+        // cleaning; in most cases, it will not grow much, and in others, it may
+        // actually be useful to cache more.
+        cache_size += files[i].size;
```

Comment:
I would sum first and skip all the debugging logging and stuff if it's not at the limit

(could even do this in the readdir-loop)

Reply:
I find it useful logs, not only for development, but also for logs in case someone reports issues with some insane cache contents, we would see that. And it also makes clear to user that those files are "managed" by mpv and gives easy way to check their age, without going in and checking atime manually.

Reply:
I disagree.
If I enable debug logging I don't want mpv to tell me every time about the files that have been sitting in my cache for months (*unless* it is taking action on those files). If I wanted to know what is sitting there I'd go check myself.
In 99% of cases it's just an extra 20-100 lines of clutter.

> but also for logs in case someone reports issues with some insane cache contents, we would see that

`cache_load_obj` and `cache_save_obj` already log the sizes, too.

Reply:
> 20-100 lines of clutter.

Would `MP_TRACE` fix your concern?